cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternalsCommon.qll

```diff
@@ -246,14 +246,6 @@ private module IteratorIndirections {
       baseType = super.getValueType()
     }
 
-    override predicate isAdditionalDereference(Instruction deref, Operand address) {
-      exists(CallInstruction call |
-        operandForFullyConvertedCall(getAUse(deref), call) and
-        this = call.getStaticCallTarget().getClassAndName("operator*") and
-        address = call.getThisArgumentOperand()
-      )
-    }
-
```

Comment:
At first sight this not only applies to iterators, but also in other contexts. So is it correct to remove this?

Reply:
In this case this only applies to iterators since `this instanceof Interfaces::Iterator` (in the class definition) and `this = call.getStaticCallTarget().getClassAndName("operator*")` (in this predicate)

Reply:
Ah, I see. What I don't understand in that case is why this removes some of the predicates, and not the whole class.

Reply:
I wanted to remove it, but we still need the extra indirection that's generated by extending `Indirection`. I'd like to totally remove this as a follow-up eventually, but this PR became large enough as-is 😂

Reply:
Ok, so what determined which predicates were removed and which weren't?

Reply:
Tests told me what we had to keep around. In terms of what is deleted, they make sense because:

- Modelling `operator*` as a dereference really doesn't make a lot of sense since `operator*` should return a reference and so it doesn't really dereference anything.
- Taint through `operator+=` and `operator-=` isn't needed anymore since we'll use dataflow SSA to find the iterator creation point (in https://github.com/github/codeql/pull/16345/commits/683fe260349ee5cf0277f47c6fac7e9109d6ad61) using `getAnUltimateDefinition` which ships past these operations.

Reply:
Clear, thanks for the clarifications.